proposals/jvm-meta-annotations-artifact.md

```diff
@@ -0,0 +1,160 @@
+
+## Motivation
+
+- We need to put somewhere `@ApplyToTypeArgumentsAnnotation` meta-annotation (see the [discussion](https://github.com/Kotlin/KEEP/issues/79#issuecomment-336905480)).
+- There is a [modules-related issue](https://blog.codefx.org/java/jsr-305-java-9/) with JSR-305 and Java 9.
+- It's worth simplifying the way how JSR-305 nullability meta-annotations are being used 
+and integrating them with Kotlin-specific meta annotations. Namely, `@UnderMigration` and `@ApplyToTypeArguments`.
+
+## Description
+
+This section describes proposed semantics of the new annotations and partly the layout of resulting artifact.
+
+### Root package 
+Root package for this artifact will be `kotlin.annotations.jvm`. 
+All classes/packages names are assumed to be placed in the package and only their 
+relative names are mentioned below.
+
+### Built-in qualifiers 
+In JSR-305, there is [`@TypeQualifier`](https://aalmiray.github.io/jsr-305/apidocs/javax/annotation/meta/TypeQualifier.html) 
+annotation that allows to introduce custom qualifiers (like `@Nonnull`).
+But that kind of meta-meta level seems to be unnecessary to our needs (at least for now), 
+the Kotlin compiler supports only nullability qualifier 
+(and in the nearest future mutability might be supported as well).
+
+So, there will only be fixed number of built-in qualifier annotations:
+- `nullability.Nullable`
+- `nullability.NotNull`
+- `mutability.Mutable`
+- `mutability.ReadOnly`
+
+Their target set would be the following:
+ElementType.METHOD, ElementType.FIELD, 
+ElementType.PARAMETER, ElementType.LOCAL_VARIABLE
+
+And semantics when being applied to types is just the same as for analogue 
+from `org.jetbrains.annotations` (see [more](https://github.com/JetBrains/kotlin/blob/master/spec-docs/flexible-java-types.md#more-precise-type-information-from-annotations) about types enhancement)
+
+### Alias annotation
+This proposal suggests to introduce `meta.Alias` annotation which would affect the compiler
+in a similar way that `@TypeQualifierNickname` [does](https://github.com/Kotlin/KEEP/blob/master/proposals/jsr-305-custom-nullability-qualifiers.md#type-qualifier-nicknames).
+
+In a basic version its declaration may look like
+```kotlin
+package kotlin.annotations.jvm.meta
+
+@Target(AnnotationTarget.ANNOTATION_CLASS)
+annotation class Alias(val qualifier: KClass<*>)
+```
+
+and its usages is supposed to look like:
+```kotlin
+@Alias(kotlin.annotations.jvm.nullability.Nullable::class)
+annotation class MyNullable
+```
+
+Thus, applying `@MyNullable` should have the same effect as `@Nullable` itself has.
+Beside it, `@MyNullable` may have `@UnderMigration` annotation on it that would change its 
+[migration status](https://github.com/Kotlin/KEEP/blob/master/proposals/jsr-305-custom-nullability-qualifiers.md#undermigration-annotation).
+
+### Default qualifiers options<a name="default-qualifiers"></a>
+This proposal has two options how
```

Comment:
I think we had three:

1. Alias, ApplyByDefault, UnderMigration
2. Alias(propagateTo), UnderMigration
3. Alias(propagateTo, migrationStatus)

Reply:
Your 2d and 3d options are different only by the way how the migration phase is defined.
And this is a different issue described in the next section.